File: cpp/102.Binary Tree Level Order Traversal.cpp
/**
 * Source: LeetCode  102. Binary Tree Level Order Traversal
 * Author: ceezyyy
 * Date: 10-06-2019
 */


 /*
 Medium

 Given a binary tree, return the level order traversal of its nodes' values.
 (ie, from left to right, level by level).

 For example:

 Given binary tree [3,9,20,null,null,15,7],

	3
   / \
  9  20
	/  \
   15   7

 return its level order traversal as:

 [
  [3],
  [9,20],
  [15,7]
 ]

 */


#include<vector>
using namespace std;

 /**
  * Definition for a binary tree node.
  * struct TreeNode {
  *     int val;
  *     TreeNode *left;
  *     TreeNode *right;
  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  * };
  */
class Solution {
public:
	struct TreeNode {
		int val;
		TreeNode *left;
		TreeNode *right;
		TreeNode(int x) : val(x), left(NULL), right(NULL) {}
	};
	vector<vector<int>> levelOrder(TreeNode* root) {
		vector<vector<int>> res;
		queue<TreeNode*> q;
		if (root == NULL)
			return res;
		q.push(root);  // enter the queue
		while (!q.empty()) {
			vector<int> temp;  // subarray, each level
			int size = q.size();
			// make sure to return the level order traversal
			while (size--) {
				TreeNode* node = q.front();
				q.pop();
				if (node->left)
					q.push(node->left);
				if (node->right)
					q.push(node->right);
				temp.push_back(node->val);
			}
			res.push_back(temp);  // each level

		}
		return res;
	}
};


/**
 * Runtime: 8 ms, faster than 57.14% of C++ online submissions for Binary Tree Level Order Traversal.
 * Memory Usage: 13.7 MB, less than 98.59% of C++ online submissions for Binary Tree Level Order Traversal.
 */

File: Others/Python3/102.Binary Tree Level Order Traversal.py
"""
@Title: 102. Binary Tree Level Order Traversal
@Tag: binary tree
@Date: Mar-01 2020
@Author: ceezyyy
@Difficulty: Medium
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:  # corner case
            return []
        res, queue = [], [root]
        while queue:  # "still waiting in line"
            level, n = [], len(queue)
            for i in range(n):
                # pop the first element in the queue each time
                node = queue.pop(0)
                level.append(node.val)
                if node.left:  # check left child
                    queue.append(node.left)
                if node.right:  # check right child
                    queue.append(node.right)
            res.append(level)
        return res


"""
Runtime: 28 ms, faster than 91.54% of Python3 online submissions for Binary Tree Level Order Traversal.
Memory Usage: 13.1 MB, less than 100.00% of Python3 online submissions for Binary Tree Level Order Traversal.
"""


"""
Time Complexity: O(n)
Space Complexity: O(1)
"""
